### PR TITLE
Refactor WASM input and dom-callbacks to work with multithreading

### PR DESCRIPTION
**Pull request by maxkatz6**
## What does the pull request do?

Continuing https://github.com/AvaloniaUI/Avalonia/pull/15709

Some points about WASM .NET multithreading and Avalonia after this PR:
1. Pointer/keyboard input now works. For the most part we always assume that all events are handled.
2. Drag/drop only works in single-threaded mode. 
3. Resizing works again, but very junky with multithreading.
4. Browser IME is semi-working, but it's not a regression aparently.
5. Various DOM callbacks are now working in multithreading with this PR.
6. File dialogs are still broken with multithreading.

Additionally, input handling should be generally faster now on single threaded platform too.

### CODE REVIEW

**Comment by maxkatz6** [comment]
Marking as DRAFT, as I need to double/triple check if anything was broken on .NET 8 single threaded. Will do it tomorrow/after tomorrow.

File: src/Browser/Avalonia.Browser/BrowserInputHandler.cs [hunk 1]
```diff
@@ -214,7 +178,7 @@ public bool OnDragEvent(JSObject args)
         _ = AvaloniaModule.ImportStorage();
 
         var position = new Point(args.GetPropertyAsDouble("offsetX"), args.GetPropertyAsDouble("offsetY"));
-        var modifiers = GetModifiers(args);
+        var modifiers = RawInputModifiers.None;// GetModifiers(args);
```
Comment (maxkatz6) [inline src/Browser/Avalonia.Browser/BrowserInputHandler.cs:181]: nit: fix it before merging

File: src/Browser/Avalonia.Browser/webapp/build.js [hunk 2]
```diff
@@ -8,7 +8,7 @@ require("esbuild").build({
     bundle: true,
     minify: true,
     format: "esm",
-    target: "es2018",
+    target: "es2019",
```
Comment (maxkatz6) [inline src/Browser/Avalonia.Browser/webapp/build.js:11]: flatMap usage requires es2019.
At this point we can bump it a little.

File: src/Browser/Avalonia.Browser/BrowserAppBuilder.cs [hunk 3]
```diff
@@ -78,8 +87,35 @@ public static class BrowserAppBuilder
             .AfterApplicationSetup(_ =>
             {
                 lifetime.View = new AvaloniaView(mainDivId);
-            })
-            .SetupWithLifetime(lifetime);
+            });
+
+        if (BrowserWindowingPlatform.IsManagedDispatcherEnabled)
```
Comment (kekekeks) [inline src/Browser/Avalonia.Browser/BrowserAppBuilder.cs:92]: I suspect that  `AvaloniaLocator.Current.GetService<BrowserPlatformOptions>()?.PreferManagedThreadDispatcher != false;` check will always succeed regardless of the `options` parameter passed to StartBrowserAppAsync. Only one registered with `.With` will be used

File: src/Browser/Avalonia.Browser/Interop/CanvasHelper.cs [hunk 4]
```diff
@@ -1,25 +1,23 @@
...
 internal static partial class CanvasHelper
 {
-    [JSImport("CanvasSurface.onSizeChanged", AvaloniaModule.MainModuleName)]
-    public static partial void OnSizeChanged(
-        JSObject canvasSurface,
-        [JSMarshalAs<JSType.Function<JSType.Number, JSType.Number, JSType.Number>>]
-        // TODO: this callback should be <int, int, double>. Revert after next .NET 9 preview.  
-        Action<double, double, double> onSizeChanged);
+    [JSExport]
+    public static Task OnSizeChanged(int topLevelId, double width, double height, double dpr)
```
Comment (kekekeks) [inline src/Browser/Avalonia.Browser/Interop/CanvasHelper.cs:12]: I think that we are supposed to dispatch this synchronously in ST mode, maybe via Dispatcher.Invoke

File: src/Browser/Avalonia.Browser/Interop/NavigationHelper.cs [hunk 5]
```diff
@@ -8,6 +10,13 @@ internal static partial class NavigationHelper
     [JSImport("NavigationHelper.addBackHandler", AvaloniaModule.MainModuleName)]
     public static partial void AddBackHandler([JSMarshalAs<JSType.Function<JSType.Boolean>>] Func<bool> backHandlerCallback);
 
+    public static Task<bool> OnBackRequested()
```
Comment (kekekeks) [inline src/Browser/Avalonia.Browser/Interop/NavigationHelper.cs:13]: Probably need to use dispatcher for managed dispatcher mode.

File: src/Browser/Avalonia.Browser/Interop/InputHelper.cs [hunk 6]
```diff
@@ -1,75 +1,90 @@
...
-        Func<JSObject, bool> pointerDown,
-        [JSMarshalAs<JSType.Function<JSType.Object, JSType.Boolean>>]
-        Func<JSObject, bool> pointerUp,
-        [JSMarshalAs<JSType.Function<JSType.Object, JSType.Boolean>>]
-        Func<JSObject, bool> pointerCancel,
-        [JSMarshalAs<JSType.Function<JSType.Object, JSType.Boolean>>]
-        Func<JSObject, bool> wheel);
+    public static Task RedirectInputAsync(int topLevelId, Action<BrowserTopLevelImpl> handler)
+    {
+        // TODO consider passing input events through dispatcher or something.
```
Comment (kekekeks) [inline src/Browser/Avalonia.Browser/Interop/InputHelper.cs:11]: Probably an outdated comment, since InputHandler uses dispatcher internally when needed

**Review by kekekeks — APPROVED** [review]
LGTM in general.


We probably need to auto-register options passed to the start method in the locator.

File: src/Browser/Avalonia.Browser/Interop/InputHelper.cs (hunk 6 above)
Comment (maxkatz6) [reply src/Browser/Avalonia.Browser/Interop/InputHelper.cs:12]: Yea

File: src/Browser/Avalonia.Browser/BrowserAppBuilder.cs (hunk 3 above)
Comment (maxkatz6) [reply src/Browser/Avalonia.Browser/BrowserAppBuilder.cs:92]: We already have `AvaloniaLocator.CurrentMutable.Bind<BrowserPlatformOptions>().ToConstant(options);` for that case. It's auto-registered.

File: src/Browser/Avalonia.Browser/Interop/CanvasHelper.cs (hunk 4 above)
Comment (maxkatz6) [reply src/Browser/Avalonia.Browser/Interop/CanvasHelper.cs:12]: +

**Pull request merged**